### PR TITLE
(QENG-3370) Add pe_* value CLI override options.

### PR DESCRIPTION
**Pull request by waynr**

### CODE REVIEW

**Comment by waynr** [comment]
@doug-rosser how does this look?

Also, please note that this patch still needs tests added.

File: test/test_generator.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  def test_fixtures
+    Find.find 'test/fixtures' do |f|
+      if File.directory?(f)
+        next
+      end
+
+      fixture_file = File.open(f, "r")
+      fixture_hash = YAML.load(fixture_file)
```
Comment (rick) [inline test/test_generator.rb]: Idiomatically, I would probably drop the temp file variable and go with:

``` ruby
fixture_hash = YAML.load_file(f)
```

or

``` ruby
fixture_hash = YAML.load(File.read(f))
```

(The "or" because I can't remember if `load_file` is always available, may depend on YAML version or implementation).

**Comment by rick** [comment]
OK, so the fixtures are used to snapshot the behavior of the generator at this point in time?

File: test/test_generator.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+      options = fixture_hash["options_string"]
+      options = options.split
+      test_hash = run_cli_with_options(options)
+      assert_equal(fixture_hash["expected_hash"], test_hash)
```
Comment (rick) [inline test/test_generator.rb]: This is probably a good place to document (for instance via the 3rd argument to `assert_equal`) what this failure means. I believe it means "the behavior of the generator has changed from the snapshotted behavior".

If that's true folks are also going to want guidance on how/when to re-snapshot things so that they can take the changed behavior as a new baseline.

**Comment by waynr** [comment]
@rick yeah, snapshotting is a good way to describe it. I think I've addressed all your comments. I added a `generate:fixtures` rake task and added an `assert_equal` mismatch helper message indicating which fixture borked and the options passed to `beaker-hostgenerator`.

Unfortunately the tests stop running on the first assertion failure so I'm still not sure this is a particularly good way to write the `test_fixtures` function. It'd be nice to see all the broken fixtures for a given `beaker-hostgenerator` change at once.

**Comment by doug-rosser** [comment]
@waynr The code looks good, but the old (weird) behavior is not present. Even though I think the usefulness is dubious, the old code (master branch) would set pe_dir if pe_family and pe_version were the same string. This new code doesn't preserve that behavior. Example:
`[beaker-hostgenerator (master)]$ env | grep pe_
pe_family=2016.1.0
pe_version=2016.1.0
[beaker-hostgenerator (master)]$ bin/beaker-hostgenerator centos6-64mdca
WARNING: Starting with beaker-hostgenerator 1.x platform strings for "el" hosts
will correspond to the actual linux distribution name. ie, the platform string
corresponding to a host specified as "centos4_64a" will be "centos-4-x86_64"
rather than "el-4-x86_64". It is recommended that you update your project's test
suites ASAP or be forced to do so when beaker-hostgenerator development moves on
to the 1.x series. We don't intend to backport features or platforms to 0.x.
---
HOSTS:
  centos6-64-1:
    pe_dir: http://enterprise.delivery.puppetlabs.net/archives/releases/2016.1.0/
    pe_ver: 2016.1.0
    pe_upgrade_dir: 
    pe_upgrade_ver: 
    hypervisor: vmpooler
    platform: el-6-x86_64
    template: centos-6-x86_64
    roles:
    - agent
    - master
    - database
    - dashboard
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
[beaker-hostgenerator (master)]$ 
[beaker-hostgenerator (master)]$ git checkout qeng-3370 
Switched to branch 'qeng-3370'
Your branch is up-to-date with 'upstream/qeng-3370'.
[beaker-hostgenerator (qeng-3370)]$ 
[beaker-hostgenerator (qeng-3370)]$ bin/beaker-hostgenerator centos6-64mdca
WARNING: Starting with beaker-hostgenerator 1.x platform strings for "el" hosts
will correspond to the actual linux distribution name. ie, the platform string
corresponding to a host specified as "centos4_64a" will be "centos-4-x86_64"
rather than "el-4-x86_64". It is recommended that you update your project's test
suites ASAP or be forced to do so when beaker-hostgenerator development moves on
to the 1.x series. We don't intend to backport features or platforms to 0.x.
---
HOSTS:
  centos6-64-1:
    pe_dir: 
    pe_ver: 
    pe_upgrade_dir: 
    pe_upgrade_ver: 
    hypervisor: vmpooler
    platform: el-6-x86_64
    template: centos-6-x86_64
    roles:
    - agent
    - master
    - database
    - dashboard
CONFIG:
  nfs_server: none
  consoleport: 443
  pooling_api: http://vmpooler.delivery.puppetlabs.net/
`

**Comment by waynr** [comment]
@doug-rosser okay, I've fixed that bug. The problem was here: https://github.com/puppetlabs/beaker-hostgenerator/commit/aa3e08cdca7bc40e9eae037ac625e7c5fc6acc22#diff-0623c137fe4358c047889a4ff1adc01bL62

There were some other module constant initialization problems throughout the beaker-hostgenerator library that were leading to incorrectly-generated fixtures which have now been fixed.

**Comment by doug-rosser** [comment]
:+1: :+1: :+1:

File: .projectile [hunk 3]
```diff
@@ -0,0 +1 @@
+-/vendor
```
Comment (nwolfe) [inline .projectile]: Do we want to commit an emacs file like this?

Comment (waynr) [inline .projectile]: I added it because I think of it as similar to a `.gitignore` file which can often contain subjectively-named file paths. e.g. https://github.com/puppetlabs/beaker-hostgenerator/blob/master/.gitignore#L13-L14 (That `.gitignore` originally came from puppetlabs/beaker-template)

The difference here is that I don't care as much about `.gitignore` so I never bothered to add `vendor` to it the way I do for projectile. Local package installation is much more annoying in projectile than in git.

Comment (waynr) [inline .projectile]: I would defer to you as far as whether you think it's a good idea or not--I haven't really thought through the implications for other emacs users. I'm sure there is some way to configure projectile exclude paths in one's emacs configuration which would addres my problem more generally.

Comment (nwolfe) [inline .projectile]: I think you want to add `.projectile` to your global `~/.gitignore`, and then you can sprinkle these projectile ignore files anywhere.

Comment (nwolfe) [inline .projectile]: Alternately, you could just configure projectile to globally ignore `vendor` directories with `projectile-globally-ignored-files` if that suits you better.  

Either way I wouldn't think we'd want to start committing editor-specific files.

Comment (waynr) [inline .projectile]: Are you saying that I should remove this file from the PR? I don't have a problem with that, but just to clarify I intentially added it in the first place.

Comment (nwolfe) [inline .projectile]: Ha yeah I'd say remove it from the PR and achieve what you're trying to do using one of the options I mentioned above.

But this is a good idea, I'm going to ignore vendor directories too :smile:

File: README.md [hunk 4]
```diff
@@ -86,6 +101,82 @@ CONFIG:
...
+likely made a change that incompatibly changes the behavior of
+`beaker-hostgenerator` for other users. Use the test fixtures as a guide to
+figure out what you did wrong and figure out how to achieve your goal without
+potentially breaking `beaker-hostgenerator` for other users.
+
+There are are few circumstances when you should expect to run the
+`generate:fixtures` task:
+
+- When you modify the `FixtureGenerator` to generate new fixtures.
+- When you need to fix bug (generated hosts are not usable without your change,
```
Comment (rick) [inline README.md]: :lipstick: "fix a bug"

Comment (waynr) [inline README.md]: Fixed!

File: Gemfile [hunk 5]
```diff
@@ -21,6 +21,8 @@ group :acceptance_testing do
   gem "beaker", *location_for(ENV['BEAKER_VERSION'] || '~> 2.0')
 end
 
+gem "minitest"
```
Comment (nwolfe) [inline Gemfile]: Should this be under a test or development group so it's not required for normal usage?

File: README.md [hunk 6]
```diff
@@ -86,6 +101,82 @@ CONFIG:
...
+task.
+
+However, this is not something that should need to be done very often. If you
+are running tests and find that some fixtures no longer work, you have most
+likely made a change that incompatibly changes the behavior of
+`beaker-hostgenerator` for other users. Use the test fixtures as a guide to
+figure out what you did wrong and figure out how to achieve your goal without
+potentially breaking `beaker-hostgenerator` for other users.
+
+There are are few circumstances when you should expect to run the
```
Comment (nwolfe) [inline README.md]: s/are are/are

File: test/util/generator_helpers.rb [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+require 'beaker-hostgenerator'
+require 'beaker-hostgenerator/data/vmpooler'
+require 'beaker-hostgenerator/data'
+
+module GeneratorTestHelpers
+  include BeakerHostGenerator::Data
+  include BeakerHostGenerator::Data::Vmpooler
+
+  def run_cli_with_options options = []
```
Comment (nwolfe) [inline test/util/generator_helpers.rb]: We should put the parens around the method parameters here and `open_file` below, for readability and consistency

File: test/util/generator_helpers.rb [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+    STDERR.reopen("stderr.txt", "w")
+    cli = BeakerHostGenerator::CLI.new(options)
+    yaml_string = cli.execute
+
+    hash = YAML.load(yaml_string)
+    return hash
```
Comment (nwolfe) [inline test/util/generator_helpers.rb]: Whoa, forgot this was a keyword! Looks like most code doesn't bother using `return`, I'd probably remove it but no big deal

File: test/util/generator_helpers.rb [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  def generate
+    # Validates single-host scenarios using all short-form role aliases with no
+    # optional flags'
+    generate_fixtures_using_osinfo(['default'], @simple_roles.cycle, [])
+
+    # Validates the use of environment variables to set various pe options.
+    [
+      {'case_name' => 'pe_version_and_pe_family',
+       'environment_variables' => {
+         'pe_version' => '6.6.6',
```
Comment (nwolfe) [inline test/util/generator_helpers.rb]: :metal:

**Comment by nwolfe** [comment]
:+1: - left some minor nitpicks

File: test/util/generator_helpers.rb (hunk 8 above)
Comment (waynr) [inline test/util/generator_helpers.rb]: I would prefer to leave it in and deal with it in a more comprehensive rubocop-driven style refactor. I guess the implied return value of the last expression in the function is probably pretty familiar for you coming from clojure and all...I happen to prefer explicit over implicit for behaviors like this (coming from C/bash/python world) but will definitely be changing this in a later cleanup PR.

File: Rakefile [hunk 10]
```diff
@@ -37,10 +48,25 @@ A quick acceptance test, named because it has no pre-suites to run
...
+    fixgen = FixtureGenerator.new
+    fixgen.generate
+  end
+end
+
 # namespace-named default tasks.
 # these are the default tasks invoked when only the namespace is referenced.
 # they're needed because `task :default` in those blocks doesn't work as expected.
-task 'test:spec' => 'test:spec:run'
+task 'test:spec' => ['test:spec:run', 'test:spec:minitest']
```
Comment (er0ck) [inline Rakefile]: shouldn't :generate:fixtures be a dependency of this task?

File: lib/beaker-hostgenerator/data.rb [hunk 11]
```diff
@@ -1,11 +1,21 @@
...
+      ENV['pe_family']
+    end
+
+    def pe_upgrade_version
+      ENV['pe_upgrade_version']
+    end
+
+    def pe_upgrade_family
+      ENV['pe_upgrade_family']
+    end
```
Comment (er0ck) [inline lib/beaker-hostgenerator/data.rb]: any reason these are lowercase?
also, beaker prepends these with BEAKER_
should we do something similar here?

File: lib/beaker-hostgenerator/data.rb [hunk 12]
```diff
@@ -51,5 +61,7 @@ module Data
         'consoleport' => 443,
       }
     }
+
+    module_function :pe_version, :pe_family, :pe_upgrade_version, :pe_upgrade_family
```
Comment (er0ck) [inline lib/beaker-hostgenerator/data.rb]: the only place i see this module used is: test/util/generator_helpers.rb
why not just mix this module in?  particularly if the above file ends up with multiple modules/classes.

File: lib/beaker-hostgenerator/util.rb [hunk 13]
```diff
@@ -22,7 +22,7 @@ def pe_dir(version, family)
 
     def fixup_node(cfg)
       # PE 2.8 doesn't exist for EL 4. We use 2.0 instead.
-      if cfg['platform'] =~ /el-4/ and PE_VERSION =~ /2\.8/
+      if cfg['platform'] =~ /el-4/ and BeakerHostGenerator::Data.pe_version =~ /2\.8/
```
Comment (er0ck) [inline lib/beaker-hostgenerator/util.rb]: should this kind of biz data go in another, single place?

File: test/fixtures/default/arista4-32a [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+      hypervisor: vmpooler
+      platform: eos-4-i386
+      template: arista-4-i386
+      roles:
+      - agent
+  CONFIG:
+    nfs_server: none
+    consoleport: 443
+    pooling_api: http://vmpooler.delivery.puppetlabs.net/
+expected_exception:
```
Comment (er0ck) [inline test/fixtures/default/arista4-32a]: should all these fixtures be checked in?  are these not what are being generated by the code above?

File: Rakefile (hunk 10 above)
Comment (waynr) [inline Rakefile]: No, see the documentation at:
https://github.com/puppetlabs/beaker-hostgenerator/blob/qeng-3370/README.md#generated-fixtures

The fixture generator is not meant to be run every time the tests are run--that's why the fixtures are committed to the repo, to "snapshot" the current behavior of beaker-hostgenerator in a way that helps prevent regressions as we refactor or repurpose parts of the code.

Anyway, we generated fixtures before every time we run the specs that use those fixtures that spec would always pass unless we do something complicated to generate the fixtures for some older commit of the repo then test using them against the HEAD of the PR branch (which would assume a certain CI setup).

File: lib/beaker-hostgenerator/data.rb (hunk 11 above)
Comment (waynr) [inline lib/beaker-hostgenerator/data.rb]: Historicaly reasons mostly:
https://github.com/puppetlabs/sqa-utils-gem/blob/master/bin/genconfig#L18-L22

I'm definitely willing to consider changing their names for beaker-hostgenerator 1.x but doing so will have implications throughout CI and probably user scripts.

File: lib/beaker-hostgenerator/data.rb (hunk 12 above)
Comment (waynr) [inline lib/beaker-hostgenerator/data.rb]: I don't know, nothing in Ruby makes sense to me and whenever I think too hard about it I become immaturely angry with myself for working a job that requires me to write a language that I consider to be bad programming practice. (Things I tend to think of as good programming practice often turn out to be "ineloquent" in the duby community.)

On a more reasonable note, I may have done this back when I tried calling these functions when setting values for a module constant here:
https://github.com/puppetlabs/beaker-hostgenerator/blob/master/lib/beaker-hostgenerator/generator.rb#L14

As it turns out, the use of a module constant whose value ultimately is set programatically and can change in different circumstances is actually quite a bad idea (hopefully I've learned my lesson). If I dwell too much on this point I am fairly certain I will become irrationally angry that ruby allows module constants to be set with functions or even to values that can be mutated at all.

TL;DR I think the use of mixins is reasonable so I'll go ahead and do that, thanks for the feedback!

File: lib/beaker-hostgenerator/util.rb (hunk 13 above)
Comment (waynr) [inline lib/beaker-hostgenerator/util.rb]: Yes, but changing it will require a major version bump. (It is supported here for backwards compatibility with https://github.com/puppetlabs/sqa-utils-gem/blob/master/bin/genconfig#L259-L264)

Comment (waynr) [inline lib/beaker-hostgenerator/util.rb]: Unless you were suggesting that it go elsewhere in the beaker-hostgenerator library (my above comment assumes you meant that it should be somewhere outside this repo).

File: lib/beaker-hostgenerator/data.rb (hunk 12 above)
Comment (waynr) [inline lib/beaker-hostgenerator/data.rb]: This has been fixed.

File: lib/beaker-hostgenerator/data.rb [hunk 15]
```diff
@@ -1,11 +1,21 @@
 module BeakerHostGenerator
   module Data
 
-    # Pull various informations out of the environment.
-    PE_VERSION=ENV['pe_version']
-    PE_FAMILY=ENV['pe_family']
-    PE_UPGRADE_VERSION=ENV['pe_upgrade_version']
-    PE_UPGRADE_FAMILY=ENV['pe_upgrade_family']
+    def pe_version
+      ENV['pe_version']
```
Comment (justinstoller) [inline lib/beaker-hostgenerator/data.rb]: Do we want to memoize calls to the environment here or do we expect these values to change?

Comment (waynr) [inline lib/beaker-hostgenerator/data.rb]: We expect these values to change in the context of both test fixture generation and test fixture validation:
https://github.com/puppetlabs/beaker-hostgenerator/blob/qeng-3370/spec/beaker-hostgenerator/generator_spec.rb#L106-L113
https://github.com/puppetlabs/beaker-hostgenerator/blob/qeng-3370/test/util/generator_helpers.rb#L34-L40

Otherwise I probably would probably hate Ruby slightly less right now.

**Comment by justinstoller** [comment]
In general this looks like a good change and it has great documentation and automation around it.

However, I usually consider fixture/data driven testing like this an anti-pattern. I think it has a valuable place when you have legacy behavior around spaghetti code that is hard to test otherwise, or you have folks writing tests that may not be familiar with the code under test and/or development in general. I would prefer testing the _logic_ in the code not the variations in the data.

Having said that, that's the only thing I have an issue with, and it would appear from the previous comments that I'm in the minority, just felt obligated to mention it.

All in all, :+1:

**Comment by er0ck** [comment]
justin has a good point about what appears to be feature level testing.  i'm okay with that. maybe should move to an integration suite, and unit tests should be added for unit/class/module logic
